### PR TITLE
Manual cherrypick of #65034 to 1.9: make json serializer case sensitive

### PR DESCRIPTION
**Pull request by nikhita**
fixes partially https://github.com/kubernetes/kubernetes/issues/64612

This PR imports the latest jsoniterator library so that case sensitivity during unmarshalling is optional. The PR also sets Kubernetes json serializer to be case sensitive.

**Release note**:

```release-note
ACTION REQUIRED: Kubernetes JSON deserializer is now case-sensitive to restore compatibility with pre-1.8 servers. If your config files contains fields with wrong case, the config files will be now invalid.
```

/sig api-machinery
/kind bug
/assign caesarxuchao liggitt thockin sttts mbohlool

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/group_version_test.go [hunk 1]
```diff
@@ -21,7 +21,7 @@ import (
 	"reflect"
 	"testing"
 
-	jsoniter "github.com/json-iterator/go"
+	k8s_json "k8s.io/apimachinery/pkg/runtime/serializer/json"
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/group_version_test.go]: what an ugly alias. Did we have that as well in the original PR?

**Comment by sttts** [comment]
@nikhita this was a clean pick?

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/group_version_test.go (hunk 1 above)
Comment (nikhita) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/group_version_test.go]: > Did we have that as well in the original PR?

Yes.

https://github.com/kubernetes/kubernetes/pull/65034/files#diff-32a177aa7704bf6900c86916b79541c3R24

**Comment by nikhita** [comment]
> @nikhita this was a clean pick?

No, there were conflicts but it was mostly changing the `Unmarshal` lines, like this: https://github.com/kubernetes/kubernetes/pull/65298/commits/00c19852bfe020ee4d5b1e2043522596cde4211f#diff-7ed5a29e1e19e4b00f6729f6bf6eadc2R62

**Comment by sttts** [comment]
> No, there were conflicts

ok, it sgtm, but leaving the final lgtm to @caesarxuchao

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types_test.go [hunk 2]
```diff
@@ -45,7 +45,7 @@ func TestVerbsUgorjiMarshalJSON(t *testing.T) {
 	}
 }
 
-func TestVerbsUgorjiUnmarshalJSON(t *testing.T) {
+func TestVerbsUJsonIterUnmarshalJSON(t *testing.T) {
```
Comment (nikhita) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types_test.go]: nit: looks like we have an extra U here :eyes:

File: Godeps/Godeps.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
 	"ImportPath": "k8s.io/kubernetes",
-	"GoVersion": "go1.9",
+	"GoVersion": "go1.10",
```
Comment (liggitt) [inline Godeps/Godeps.json]: this change doesn't look correct

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types_test.go (hunk 2 above)
Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types_test.go]: also in master

File: Godeps/Godeps.json (hunk 3 above)
Comment (nikhita) [reply Godeps/Godeps.json]: fixed

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types_test.go (hunk 2 above)
Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types_test.go]: no, don't fix it here

**Comment by nikhita** [comment]
/priority critical-urgent

Can someone add the milestone?

**Comment by nikhita** [comment]
Opened https://github.com/kubernetes/kubernetes/pull/65324 to fix nits on master.

**Comment by mbohlool** [comment]
I just caught up on the original PR. I feel you guys want to make sure less people would depend on this. That's why you want to cherrypick it in 1.9 and 1.8. But did we also consider we broke a bunch of users on a patch release? Is there any difference in breaking people on patch release vs major releases?

**Comment by mbohlool** [comment]
@liggitt @lavalamp please let me know what you think about patching this vs leaving it for main releases only.

**Comment by lavalamp** [comment]
The risk @mbohlool is talking about is clients who have written incorrectly cased config files (stored in the users SCM or whatever, not those stored in etcd) suddenly having those break. I think we should be allowed to fix horrible bugs even if that requires user action. I double checked with @thockin and he had the same opinion.

That said, we do need to make sure this shows up in the release notes.

@liggitt Any concerns from you?

**Comment by liggitt** [comment]
> The risk @mbohlool is talking about is clients who have written incorrectly cased config files (stored in the users SCM or whatever, not those stored in etcd) suddenly having those break. I think we should be allowed to fix horrible bugs even if that requires user action. I double checked with @thockin and he had the same opinion.
> 
> That said, we do need to make sure this shows up in the release notes.
> 
> @liggitt Any concerns from you?

I agree on all counts. The interoperability concerns with such manifests are many. They would already cause problems for kubectl (both because of client side validation and stored apply annotations containing the wrong case), as well as tools that embed kubectl. 

Clients and manifests written against our published/documented schemas, or using our typed client-go libraries, would not be affected by this fix.

**Comment by mbohlool** [comment]
I am happy as long as it discussed. Thanks @lavalamp @liggitt. Cherrypick is approved.

**Comment by mbohlool** [comment]
@lavalamp The PR has a release note that will be included in 1.9.9 release with a clear Action Item for users to fix their local config files.

**Comment by caesarxuchao** [comment]
/lgtm

Thanks all.

**Comment by liggitt** [comment]
/hold
for fix to scheduler config (https://github.com/kubernetes/kubernetes/pull/65415)

**Comment by mbohlool** [comment]
@liggitt already approved #65418, do you want to unhold this? The deadline of next patch release is today.

**Comment by liggitt** [comment]
as long as you're aware #65418 is required for release, we can unhold this

/hold cancel

**Pull request merged**